サインアップ画面、ログイン画面、パスワード変更画面、設定画面

ri0007 opened the pull request:
サインアップ画面、ログイン画面、パスワード変更画面、設定画面を実装しました
デバッグの関係上、設定画面から起動します

あとFiles changedではFBSDKとParseSDKのファイルが非常に多いので自分が書いたファイルを記しておきます

AppDelegate.swift
InstagramAppExtensions.swift (UIImageViewに画像を非同期で取得＆表示させるメソッドを追加)
ParseAccess.swift (Parseにアクセスするためのラッパークラス)

TriangleView.swift (NonLoginViewControllerの「Sign Up」と「Login」のボタンを指すための三角形のビュー)

SignUpViewController.swift (サインアップ画面)
NonLoginViewController.swift (サインアップのメールアドレス入力とログインを行う本来の初期画面)
SettingsTableViewController.swift (設定画面 現在はログアウトとパスワード変更しかない)
PasswordResetViewController.swift (パスワードリセット画面)

Commit:
Parseから提供されているフレームワークと、それと依存関係にあるフレームワーク群を追加。
ViewController.swiftにParseのテストコードを記述。

Commit:
FacebookSDKの追加とそれに伴うinfo.plistの変更
Bridging-Header.hの追加及びBuild Settingsの設定
AppDelegate.swiftにFacebookのトラッキング用コードを追加

Commit:
Parseアクセス用ラッパークラスを作成(ParseAccess.swift)
サインアップメソッドの実装
SignUpViewControllerの作成

Commit:
Facebookのパーミッション取得のため、info.plistにLSApplicationQueriesSchemesを追加。
InstagramAppExtensions.swift,NonLoginViewController.swiftを追加。
Login.storyboardにNonLoginViewControllerを追加。
ParseAccess.swiftのcreateUserメソッドの名称をsignUpUserに変更。また、loginUserメソッドを追加。
SignUpViewControllerにてアイコン画像の選択(Facebookのアイコンもしくはカメラロール)を実装。

Commit:
NonLoginViewControllerのUIを作成
NonLoginViewControllerでのログインを実装
Storyboardの各UIImageViewにshibuya.jpgを設定

Commit:
SettingsTableViewController及びPasswordResetViewControllerを作成
Settings.storyboard及びPasswordReset.storyboardを作成
ParseAccessにresetPasswordメソッドを追加
設定画面からのパスワードリセットを実装

Commit:
ParseFacebookUtilsをリファレンスから削除
ParseAccessにPFFacebookUtilsを使用したsignUpUserWithFacebookメソッドを追加
NonLoginViewControllerのdidPushedFBLoginButtonメソッドで行うログインをFBSDKLoginManagerを使用したログインからParseAccessのsignUpUserWithFacebookを使用したログインに変更

Commit:
Login.storyboardの語句を統一し、NonLoginViewControllerのアクションの関連付けを変更
NonLoginViewControllerにて一部TextFieldがキーボードサイズによっては選択できない問題を修正

keitin commented on InstagramApp/InstagramAppExtensions.swift:
```diff
@@ -0,0 +1,39 @@
+
+extension UIImageView{
+    //画像を非同期で取得＆表示させる
+    func getUrlImageAsynchronous(urlString:String, complitionHander:((succeeded:Bool, error:NSError?) -> Void)?){
```
Alamofireを使って取得すると良いかな？
http://nukenuke.hatenablog.com/entry/2015/09/18/083000

ri0007 replied:
研修アプリという性質上、外部ライブラリは使わずに出来るだけ自分でコードを書いていく感じで進めていましたが入れてしまって良いですかね？
良いならFBSDKと一緒にCocoaPodsで入れます

keitin replied:
今のメンターは逆にライブラリを使う機会が少ないので
むしろばんばん使って欲しいです!!

常にベストプラクティスで実装してしまってOKです！

yuta-t replied:
コロンのあとにスペース入れましょう。

yuta-t replied:
コロンの後とブレースの前にスペース入れましょう。

```
func getUrlImageAsynchronous(urlString: String, complitionHander: ((succeeded: Bool, error: NSError?) -> Void)?) {
```

keitin commented on FBSDKCoreKit.framework/Headers/FBSDKAccessToken.h:
```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
```
FBSDKをCocoaPodsで管理できないかな？

keitin commented on InstagramApp/NonLoginViewController.swift:
```diff
@@ -0,0 +1,265 @@
+    //MARK: - ログイン関連
+    
+    @IBAction func login(sender: UIButton) {    //実際にログインする
+        let alert = UIAlertController(title: "エラー", message: nil, preferredStyle: UIAlertControllerStyle.Alert)
```
UIAlertControllerStyleの部分は省略しましょう！

ri0007 replied:
これ省略できましたっけ…？

yuta-t replied:
引数の型がUIAlertControllerStyle型となっているので、`.Alert`と書けます。

keitin commented on InstagramApp/NonLoginViewController.swift:
```diff
@@ -0,0 +1,265 @@
+    
+    @IBAction func login(sender: UIButton) {    //実際にログインする
+        let alert = UIAlertController(title: "エラー", message: nil, preferredStyle: UIAlertControllerStyle.Alert)
+        alert.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.Cancel, handler: nil))
```
UIAlertActionStyleも省略👮

keitin commented on InstagramApp/NonLoginViewController.swift:
```diff
@@ -0,0 +1,265 @@
+        let alert = UIAlertController(title: "エラー", message: nil, preferredStyle: UIAlertControllerStyle.Alert)
+        alert.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.Cancel, handler: nil))
+        
+        ParseAccess().loginUser(nameField.text, password: passField.text) { (succeeded, user, error) -> Void in
```
クロージャにおいて戻り値が無い場合は、`-> Void`の部分は省略しましょう！

keitin commented on InstagramApp/NonLoginViewController.swift:
```diff
@@ -0,0 +1,265 @@
+        alert.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.Cancel, handler: nil))
+        
+        ParseAccess().loginUser(nameField.text, password: passField.text) { (succeeded, user, error) -> Void in
+            if succeeded == true{
```
`if succeeded == true`
は
`if succeeded`で良いかな

keitin commented on InstagramApp/NonLoginViewController.swift:
```diff
@@ -0,0 +1,265 @@
+                self.presentViewController(alert, animated: true, completion: nil)
+                self.loginSucceeded()
+            }
+            else if error != nil {
```
if let 使いましょう！

すると、if文の中でerrorを毎回アンラップしなくてOKです

keitin commented on InstagramApp/NonLoginViewController.swift:
```diff
@@ -0,0 +1,265 @@
+        self.arrowView.setNeedsUpdateConstraints()
+        self.view.removeConstraint(self.arrowViewCenterIsSignUpConstraint)
+        self.view.addConstraint(self.arrowViewCenterIsLoginConstraint)
+        UIView.animateWithDuration(0.3, animations: { () -> Void in
```
`-> Void`を省略しましょう！

Commit:
コメントで指摘された"-> Void"や"== true"を削除

Commit:
https://github.com/kyohei0423/InstagramApp/pull/2#discussion-diff-42709693 の箇所を修正

Commit:
PasswordReset.storyboardのStoryboard Entry Pointを設定

Commit:
ログイン成功時にモーダルビューとして表示させているログイン関連のビューを閉じる様に修正

yuta-t commented:
VCが全体的にFatですね
ロジックはVCに属するmodelに書いて、
Viewに関する記述もVCに属するCustom Viewを作ってあげたいかなと思います。

yuta-t commented on InstagramApp/SettingsTableViewController.swift:
```diff
@@ -0,0 +1,122 @@
+            case 0:
+                let alert = UIAlertController(title: "ログアウトしてよろしいでしょうか？", message: nil, preferredStyle: UIAlertControllerStyle.Alert)
+                alert.addAction(UIAlertAction(title: "キャンセル", style: UIAlertActionStyle.Cancel, handler: nil))
+                alert.addAction(UIAlertAction(title: "ログアウト", style: UIAlertActionStyle.Default, handler: { (action) in
```
handlerのクロージャ内で引数actionは使わないので変数名を与える必要はないかと思います。

handler: { _ in self.logout() }

yuta-t commented on InstagramApp/NonLoginViewController.swift:
```diff
@@ -0,0 +1,266 @@
+                print("Cancelled")
+            }
+            else{
+                if (result.grantedPermissions.contains("public_profile") && result.grantedPermissions.contains("email")) {
```
ここはelse ifを使わないんですか？

yuta-t commented on InstagramApp/NonLoginViewController.swift:
```diff
@@ -0,0 +1,266 @@
+    func getUerDateForFB(){
+        let request=FBSDKGraphRequest(graphPath: "/me?fields=first_name,last_name,email", parameters: nil, HTTPMethod: "GET")
+        request.startWithCompletionHandler({ (connection, result, error) in
+            if(error != nil){
```
括弧が不要なので、付けない方で統一しましょう。

yuta-t commented on InstagramApp/NonLoginViewController.swift:
```diff
@@ -0,0 +1,266 @@
+            if(error != nil){
+                print(error)
+                return
+            }
```
ここそもそもguard文を使ったほうが、!=を使わないでいいですし、読みやすいと思います。

```
guard error == nil else {
  print(error)
  return
}
```

yuta-t commented on InstagramApp/NonLoginViewController.swift:
```diff
@@ -0,0 +1,266 @@
+            }
+            else{
+                print("Can't get user datas")
+            }
```
ここも例外を先にguardで書くと、!=を使わないでよかったり、例外を先にまとめられたり、インデントも深くならないので、いいと思います。

```
guard result == nil else {
  print("Can't get user datas")
  return
}
```

yuta-t commented on InstagramApp/NonLoginViewController.swift:
```diff
@@ -0,0 +1,266 @@
+    }
+    
+    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
+        if segue.identifier == "SignUp"{
```
ここもguard使うとインデント深くならないので綺麗だと思います。

```
guard segue.identifier == "SignUp" else {
  return
}
```

Commit:
UIAlertControllerStyleとUIAlertActionStyleの記述を省略

Commit:
既存のフレームワークを削除し、CocoaPodsで再度インストールした
AlamofireをCocoaPodsでインストールした
InstagramAppExtensionsのUIImageView.getUrlImageAsynchronousメソッドを使用していた部分をAlamofireで置き換えた

yuta-t commented on InstagramApp/InstagramAppExtensions.swift:
```diff
@@ -0,0 +1,39 @@
+
+import UIKit
+
+extension UIImageView{
```
ブレースの前にスペース入れましょう。

```
extension UIImageView {
```

yuta-t commented on InstagramApp/InstagramAppExtensions.swift:
```diff
@@ -0,0 +1,39 @@
+extension UIImageView{
+    //画像を非同期で取得＆表示させる
+    func getUrlImageAsynchronous(urlString:String, complitionHander:((succeeded:Bool, error:NSError?) -> Void)?){
+        let url = NSURL(string:urlString)
```
コロンのあとにスペース

yuta-t commented on InstagramApp/InstagramAppExtensions.swift:
```diff
@@ -0,0 +1,39 @@
+    //画像を非同期で取得＆表示させる
+    func getUrlImageAsynchronous(urlString:String, complitionHander:((succeeded:Bool, error:NSError?) -> Void)?){
+        let url = NSURL(string:urlString)
+        let req = NSURLRequest(URL:url!)
```
コロンのあとにスペース

yuta-t commented on InstagramApp/InstagramAppExtensions.swift:
```diff
@@ -0,0 +1,39 @@
+                // Handle error...
+                complitionHander?(succeeded: false, error: error)
+                return
+            }
```
エラーハンドリングなのでguard使いたいですね。
やはり`!=`演算子は可読性が低くなると思います。

```
guard error == nil else {
    // Handle error...
    complitionHander?(succeeded: false, error: error)
    return
}
```

yuta-t commented on InstagramApp/InstagramAppExtensions.swift:
```diff
@@ -0,0 +1,39 @@
+                return
+            }
+            
+            if let image = UIImage(data:data!){
```
コロン後とブレース前にスペース

```
if let image = UIImage(data: data!) {
```

yuta-t commented on InstagramApp/InstagramAppExtensions.swift:
```diff
@@ -0,0 +1,39 @@
+            
+            if let image = UIImage(data:data!){
+                dispatch_async(dispatch_get_main_queue(), { () in
+                    self.image=image
```
演算子の両脇にはスペース

```
self.image = image
```

yuta-t commented on InstagramApp/InstagramAppExtensions.swift:
```diff
@@ -0,0 +1,39 @@
+            else{
+                //不明なエラー
+                complitionHander?(succeeded: false, error: nil)
+            }
```
エラーはやはり全部上にまとめて書きたいので、こちらもguard使ってあげましょう。

```
guard let image = UIImage(data: data!) else {
    //不明なエラー
    complitionHander?(succeeded: false, error: nil)
    return
}

dispatch_async(dispatch_get_main_queue(), { () in
    self.image = image
    complitionHander?(succeeded: true, error: nil)
})
```

yuta-t commented on InstagramApp/ParseAccess.swift:
```diff
@@ -0,0 +1,144 @@
+import ParseFacebookUtilsV4
+import FBSDKCoreKit
+
+enum InstagramAppErrorType : Int{
```
コロン前はスペースなし、ブレース前はスペースありで

```
enum InstagramAppErrorType: Int {
```

yuta-t commented on InstagramApp/ParseAccess.swift:
```diff
@@ -0,0 +1,144 @@
+}
+
+class ParseAccess: NSObject {
+    func signUpUser(name:String?, password:String?, email:String?, icon:UIImage?, complitionHander:(succeeded:Bool, error:NSError?) -> Void){
```
コロンのあとと、ブレースの前にスペース入れましょう

```
class ParseAccess: NSObject {
+    func signUpUser(name: String?, password: String?, email: String?, icon: UIImage?, complitionHander: (succeeded: Bool, error: NSError?) -> Void) {
```

yuta-t commented on InstagramApp/ParseAccess.swift:
```diff
@@ -0,0 +1,144 @@
+
+class ParseAccess: NSObject {
+    func signUpUser(name:String?, password:String?, email:String?, icon:UIImage?, complitionHander:(succeeded:Bool, error:NSError?) -> Void){
+        let error : NSError
```
ここはなぜ`error`の定義があるのでしょうか

下で`error = NSError(domain:〜`とやってるので、そこを`let error`にすればよいかな、と

yuta-t commented:
@ri0007 
コメントしたことに関して、コメントしていない部分も見直してもらえればと思います。
あと、PRがけっこう大きいので、もう少し小さい単位で送ってください！

Commit:
guard文への変更、コロン後やブレース前等にスペースを挿入

Commit:
https://github.com/kyohei0423/InstagramApp/pull/2#discussion_r42864357 を修正

Commit:
SignupViewControllerのメソッドの一部をModelクラス及びViewクラスに移動
Login.storyboardのサインアップ画面のメインビューの関連付けをSignUpViewクラスに変更
